File: 0628-maximum-product-of-three-numbers/0628-maximum-product-of-three-numbers.py
class Solution:
    def maximumProduct(self, nums):
        # Initialize tracking elements to positive/negative infinity boundaries
        max1 = max2 = max3 = float('-inf')
        min1 = min2 = float('inf')
        
        for n in nums:
            # Shift the three largest values dynamically
            if n > max1:
                max3, max2, max1 = max2, max1, n
            elif n > max2:
                max3, max2 = max2, n
            elif n > max3:
                max3 = n
                
            # Shift the two smallest values dynamically
            if n < min1:
                min2, min1 = min1, n
            elif n < min2:
                min2 = n
                
        return max(max1 * max2 * max3, min1 * min2 * max1)

File: 0628-maximum-product-of-three-numbers/0628-maximum-product-of-three-numbers.c
#define MAX(a, b) ((a) > (b) ? (a) : (b))

int maximumProduct(int* nums, int numsSize) {
    int max1 = -1001, max2 = -1001, max3 = -1001;
    int min1 = 1001, min2 = 1001;

    for (int i = 0; i < numsSize; i++) {
        int n = nums[i];
        if (n > max1) {
            max3 = max2; max2 = max1; max1 = n;
        } else if (n > max2) {
            max3 = max2; max2 = n;
        } else if (n > max3) {
            max3 = n;
        }

        if (n < min1) {
            min2 = min1; min1 = n;
        } else if (n < min2) {
            min2 = n;
        }
    }
    return MAX(max1 * max2 * max3, min1 * min2 * max1);
}
